### ISSUE TITLE
Unreliable CPU times in torch.autograd.profiler.profile(use_cuda=True) when using CUDA

### ISSUE THREAD

**Issue by CedricPicron** [pytorch/pytorch#44784]
## Bug/Docs improvement
In https://pytorch.org/docs/stable/bottleneck.html it is highlighted that 

> Ops that do synchronize appear to be extremely expensive under regular CPU-mode profilers. In these case where timings are incorrect, the CUDA-mode autograd profiler may be helpful.

However, below experiments show that these CPU times might be unreliable altogether, even when used with `torch.autograd.profiler.profile` with `use_cuda=True`.

## To reproduce
``` python
import argparse
import copy
import torch
from torch import nn
from torch.utils._benchmark import Timer


parser = argparse.ArgumentParser()
parser.add_argument('--sync', action='store_true')
args = parser.parse_args()

# Initialization
a = torch.randn(1024, 2, 256, device='cuda')
b = torch.randn(1024, 2, 256, device='cuda')

self_attention = nn.MultiheadAttention(256, 8).to('cuda')
layers = nn.ModuleList([copy.deepcopy(self_attention) for _ in range(10)])


def run_test(a, b, layers):
    for layer in layers:
        if args.sync:
            torch.cuda.synchronize()
        a = layer(a+b, a+b, a)[0]


# Timing from torch.utils._benchmark.Timer
timer = Timer(stmt='run_test(a, b, layers)', globals=globals())
t = timer.timeit(number=3)._median*1e3

# Profiling with torch.autograd.profiler.profile(use_cuda=True)
with torch.autograd.profiler.profile(use_cuda=True) as prof:
    run_test(a, b, layers)

print(prof.table(sort_by='cpu_time', row_limit=20))
print(f"Recorded timeit time: {t: .4f} ms")
```
### Without synchronization (CPU order)
```
-----------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------------------------------------  
Name                           Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CUDA total %     CUDA total       CUDA time avg    Number of Calls  Input Shapes                                   
-----------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------------------------------------  
aten::equal                    0.07%            47.485us         7.04%            4.591ms          4.591ms          0.15%            154.816us        154.816us        1                []                                             
aten::equal                    0.07%            48.197us         6.90%            4.503ms          4.503ms          0.15%            158.527us        158.527us        1                []                                             
aten::equal                    0.08%            49.436us         6.90%            4.501ms          4.501ms          0.15%            155.424us        155.424us        1                []                                             
aten::equal                    0.07%            45.813us         6.90%            4.500ms          4.500ms          0.15%            156.031us        156.031us        1                []                                             
aten::equal                    0.07%            45.439us         6.89%            4.493ms          4.493ms          0.14%            152.828us        152.828us        1                []                                             
aten::equal                    0.08%            51.100us         6.89%            4.493ms          4.493ms          0.15%            159.584us        159.584us        1                []                                             
aten::item                     0.02%            11.760us         6.80%            4.437ms          4.437ms          0.03%            29.312us         29.312us         1                []                                             
aten::equal                    0.07%            45.903us         6.79%            4.426ms          4.426ms          0.15%            153.855us        153.855us        1                []                                             
aten::_local_scalar_dense      6.78%            4.425ms          6.78%            4.425ms          4.425ms          0.02%            22.880us         22.880us         1                []                                             
aten::equal                    0.07%            46.158us         6.75%            4.404ms          4.404ms          0.15%            157.088us        157.088us        1                []                                             
aten::equal                    0.07%            48.453us         6.71%            4.379ms          4.379ms          0.13%            140.574us        140.574us        1                []                                             
aten::item                     0.02%            11.243us         6.66%            4.341ms          4.341ms          0.03%            28.734us         28.734us         1                []                                             
aten::item                     0.02%            11.623us         6.65%            4.334ms          4.334ms          0.03%            29.217us         29.217us         1                []                                             
aten::_local_scalar_dense      6.64%            4.329ms          6.64%            4.329ms          4.329ms          0.02%            22.758us         22.758us         1                []                                             
aten::item                     0.02%            11.663us         6.64%            4.329ms          4.329ms          0.03%            29.824us         29.824us         1                []                                             
aten::item                     0.02%            12.640us         6.63%            4.323ms          4.323ms          0.03%            32.129us         32.129us         1                []                                             
aten::item                     0.02%            11.628us         6.63%            4.323ms          4.323ms          0.03%            30.500us         30.500us         1                []                                             
aten::_local_scalar_dense      6.63%            4.323ms          6.63%            4.323ms          4.323ms          0.02%            23.393us         23.393us         1                []                                             
aten::_local_scalar_dense      6.62%            4.318ms          6.62%            4.318ms          4.318ms          0.02%            24.031us         24.031us         1                []                                             
aten::_local_scalar_dense      6.61%            4.311ms          6.61%            4.311ms          4.311ms          0.02%            24.605us         24.605us         1                []                                             
-----------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------------------------------------  
Self CPU time total: 65.218ms
CUDA time total: 105.474ms
Recorded timeit time:  63.6856 ms
```

###  With synchronization (CPU order)
```
-----------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------------------------------------  
Name                           Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CUDA total %     CUDA total       CUDA time avg    Number of Calls  Input Shapes                                   
-----------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------------------------------------  
aten::matmul                   0.11%            30.299us         1.01%            275.451us        275.451us        0.36%            373.280us        373.280us        1                []                                             
aten::equal                    0.16%            43.729us         0.88%            239.862us        239.862us        0.23%            238.785us        238.785us        1                []                                             
aten::matmul                   0.13%            36.116us         0.85%            231.862us        231.862us        0.32%            339.266us        339.266us        1                []                                             
aten::equal                    0.18%            48.722us         0.83%            224.294us        224.294us        0.21%            223.230us        223.230us        1                []                                             
aten::equal                    0.17%            45.429us         0.81%            219.863us        219.863us        0.18%            191.232us        191.232us        1                []                                             
aten::equal                    0.22%            59.855us         0.80%            216.475us        216.475us        0.19%            199.898us        199.898us        1                []                                             
aten::equal                    0.19%            51.937us         0.77%            209.972us        209.972us        0.19%            194.048us        194.048us        1                []                                             
aten::equal                    0.18%            49.362us         0.77%            208.869us        208.869us        0.20%            209.055us        209.055us        1                []                                             
aten::equal                    0.17%            47.402us         0.76%            206.019us        206.019us        0.18%            190.176us        190.176us        1                []                                             
aten::equal                    0.18%            49.714us         0.75%            204.618us        204.618us        0.17%            182.336us        182.336us        1                []                                             
aten::equal                    0.19%            50.461us         0.75%            202.629us        202.629us        0.19%            198.785us        198.785us        1                []                                             
aten::equal                    0.16%            44.799us         0.74%            202.441us        202.441us        0.16%            172.512us        172.512us        1                []                                             
aten::equal                    0.18%            48.253us         0.74%            201.980us        201.980us        0.17%            174.945us        174.945us        1                []                                             
aten::matmul                   0.11%            30.177us         0.73%            199.028us        199.028us        0.19%            200.704us        200.704us        1                []                                             
aten::equal                    0.17%            46.481us         0.73%            198.270us        198.270us        0.16%            169.920us        169.920us        1                []                                             
aten::matmul                   0.12%            31.485us         0.72%            196.478us        196.478us        0.28%            297.992us        297.992us        1                []                                             
aten::matmul                   0.12%            31.645us         0.72%            196.246us        196.246us        0.29%            303.102us        303.102us        1                []                                             
aten::equal                    0.16%            43.809us         0.72%            194.799us        194.799us        0.18%            193.663us        193.663us        1                []                                             
aten::matmul                   0.11%            28.648us         0.71%            193.169us        193.169us        0.20%            204.797us        204.797us        1                []                                             
aten::matmul                   0.11%            30.296us         0.71%            193.006us        193.006us        0.29%            300.992us        300.992us        1                []                                             
-----------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------------------------------------  
Self CPU time total: 27.182ms
CUDA time total: 104.843ms
Recorded timeit time:  61.9296 ms
```

## Interpretation
One can see that the CPU times appear to change, just by synchronizing at different times. If one would synchronize at another time, again totally different CPU times might be measured. This is I believe in contrast with what is being told with

> Ops that do synchronize appear to be extremely expensive under regular CPU-mode profilers. In these case where timings are incorrect, the CUDA-mode autograd profiler may be helpful.

in https://pytorch.org/docs/stable/bottleneck.html. 

## Additional context
1. Note that the CUDA measurements are stable, regardless of synchronization. In above example, the softmax operation actually is the most expensive (not mentioned in CPU ordered tables above). Profiling based on this information seems much more reliable. 

2. A question that I have and might also be handy in profiler docs: Can the total execution be inferred from the Self CPU time total and/or from CUDA time total (sum, self cpu time only, cuda time only, not possible)? Above experiments suggest it is not possible.

cc @ngimel

**Comment by ngimel** [comment]
cc @ilia-cher

**Comment by aaronenyeshi** [comment]
There is a significant re-write to profiler, and should reduce CPU op overheads. Please check if this is still an issue
